File: packages/vibechk/src/vibechk/git.py
from __future__ import annotations

import subprocess
from abc import ABC, abstractmethod
from pathlib import Path

from vibechk.errors import VibechkError


class GitGateway(ABC):
    @abstractmethod
    def repo_root(self) -> Path:
        """Return repository root for the workdir."""

    @abstractmethod
    def current_branch(self) -> str:
        """Return current branch or fail on detached HEAD."""

    @abstractmethod
    def current_commit(self) -> str:
        """Return current commit SHA."""

    @abstractmethod
    def remotes(self) -> dict[str, str]:
        """Return remotes keyed by name."""

    @abstractmethod
    def is_clean(self) -> bool:
        """Return whether the worktree and index are clean."""

    @abstractmethod
    def diff_patch(self) -> str:
        """Return a patch for changes relative to HEAD."""

    @abstractmethod
    def has_changes(self) -> bool:
        """Return whether the worktree or index has changes."""

    @abstractmethod
    def create_result_branch_and_commit(self, branch: str, message: str) -> None:
        """Create a result branch from current HEAD and commit all changes."""

    @abstractmethod
    def checkout(self, branch: str) -> None:
        """Checkout an existing branch."""


class RealGitGateway(GitGateway):
    def __init__(self, workdir: Path) -> None:
        self._workdir = workdir

    def repo_root(self) -> Path:
        result = self._run_git_raw(["rev-parse", "--show-toplevel"])
        if result.returncode != 0:
            raise VibechkError(f"Workdir {self._workdir} is not a git repository.")
        root_text = result.stdout.strip()
        if not root_text:
            raise VibechkError(f"Workdir {self._workdir} is not a git repository.")
        return Path(root_text)

    def current_branch(self) -> str:
        branch = self._run_git(
            ["rev-parse", "--abbrev-ref", "HEAD"],
            error_message=f"Could not determine current branch in {self._workdir}",
        )
        if branch == "HEAD":
            raise VibechkError(f"Workdir {self._workdir} is in detached HEAD state.")
        return branch

    def current_commit(self) -> str:
        return self._run_git(
            ["rev-parse", "HEAD"],
            error_message=f"Could not determine current commit in {self._workdir}",
        )

    def remotes(self) -> dict[str, str]:
        output = self._run_git(
            ["remote", "-v"],
            error_message=f"Could not list git remotes in {self._workdir}",
        )
        remotes: dict[str, str] = {}
        for line in output.splitlines():
            parts = line.split()
            if len(parts) >= 2 and parts[0] not in remotes:
                remotes[parts[0]] = parts[1]
        return remotes

    def is_clean(self) -> bool:
        return self._status_porcelain() == ""

    def diff_patch(self) -> str:
        if self.has_changes():
            self._run_git(
                ["add", "-N", "."],
                error_message=f"Could not prepare untracked files for diff in {self._workdir}",
            )
        return self._run_git(
            ["diff", "--binary", "HEAD"],
            error_message=f"Could not capture git diff in {self._workdir}",
        )

    def has_changes(self) -> bool:
        return self._status_porcelain() != ""

    def create_result_branch_and_commit(self, branch: str, message: str) -> None:
        self._run_git(
            ["switch", "-c", branch],
            error_message=f"Could not create result branch {branch}",
        )
        self._run_git(
            ["add", "-A"],
            error_message=f"Could not stage vibechk changes on {branch}",
        )
        self._run_git(
            ["commit", "-m", message],
            error_message=f"Could not commit vibechk changes on {branch}",
        )

    def checkout(self, branch: str) -> None:
        self._run_git(
            ["switch", branch],
            error_message=f"Could not switch back to branch {branch}",
        )

    def _status_porcelain(self) -> str:
        return self._run_git(
            ["status", "--porcelain=v1"],
            error_message=f"Could not inspect git status in {self._workdir}",
        )

    def _run_git(self, args: list[str], *, error_message: str) -> str:
        result = self._run_git_raw(args)
        if result.returncode != 0:
            details = result.stderr.strip() or result.stdout.strip()
            if details:
                raise VibechkError(f"{error_message}: {details}")
            raise VibechkError(f"{error_message}.")
        return result.stdout.strip()

    def _run_git_raw(self, args: list[str]) -> subprocess.CompletedProcess[str]:
        try:
            return subprocess.run(
                ["git", *args],
                cwd=self._workdir,
                check=False,
                capture_output=True,
                text=True,
            )
        except FileNotFoundError as error:
            raise VibechkError("git is not installed or not on PATH.") from error

File: ts/packages/vibechk/src/git.ts
import {
	formatCommand,
	formatCommandFailure,
	NodeCommandExecApi,
	type CommandExecApi,
	type ExecResult,
} from "@asdl/core/exec";
import {
	RealGitGateway,
	type GitGateway,
	type GitOperationResult,
	type GitResult,
} from "@asdl/core/git";

import { VibechkError } from "./store.ts";

const GIT_TIMEOUT_MS = 10_000;

export interface GitProvenance {
	repoRoot: string;
	startingBranch: string;
	startingCommit: string;
	remotes: Record<string, string>;
}

export interface VibechkGitGateway {
	repoRoot(): Promise<string>;
	currentBranch(): Promise<string>;
	currentCommit(): Promise<string>;
	remotes(): Promise<Record<string, string>>;
	isClean(): Promise<boolean>;
	diffPatch(): Promise<string>;
	hasChanges(): Promise<boolean>;
	createResultBranchAndCommit(branch: string, message: string): Promise<void>;
	checkout(branch: string): Promise<void>;
}

export class RealVibechkGitGateway implements VibechkGitGateway {
	private readonly workdir: string;
	private readonly execApi: CommandExecApi;
	private readonly coreGit: GitGateway;

	constructor(workdir: string, execApi: CommandExecApi = new NodeCommandExecApi()) {
		this.workdir = workdir;
		this.execApi = execApi;
		this.coreGit = new RealGitGateway(execApi);
	}

	async repoRoot(): Promise<string> {
		const result = await this.coreGit.repoRoot({ cwd: this.workdir });
		return this.unwrapGitResult(result, `Workdir ${this.workdir} is not a git repository.`);
	}

	async currentBranch(): Promise<string> {
		const result = await this.coreGit.currentBranch({ cwd: this.workdir });
		if (result.type === "branch") return result.branch;
		if (result.type === "detached") {
			throw new VibechkError(
				`Could not determine current branch in ${this.workdir}\nDetached HEAD.`,
			);
		}
		if (isMissingExecutableError(result.error.message)) {
			throw new VibechkError("git is not installed or not on PATH.");
		}
		throw new VibechkError(
			`Could not determine current branch in ${this.workdir}\n${result.error.message}`,
		);
	}

	async currentCommit(): Promise<string> {
		const result = await this.coreGit.headCommit({ cwd: this.workdir });
		return this.unwrapGitResult(result, `Could not determine current commit in ${this.workdir}`);
	}

	async remotes(): Promise<Record<string, string>> {
		const output = await this.runGit(
			["remote", "-v"],
			`Could not list git remotes in ${this.workdir}`,
		);
		const remotes: Record<string, string> = {};
		for (const line of output.split("\n")) {
			const parts = line.split(/\s+/u);
			if (
				parts.length >= 2 &&
				parts[0] !== undefined &&
				parts[1] !== undefined &&
				!(parts[0] in remotes)
			) {
				remotes[parts[0]] = parts[1];
			}
		}
		return remotes;
	}

	async isClean(): Promise<boolean> {
		return !(await this.hasChanges());
	}

	async diffPatch(): Promise<string> {
		if (await this.hasChanges()) {
			await this.runGit(
				["add", "-N", "."],
				`Could not prepare untracked files for diff in ${this.workdir}`,
			);
		}
		return await this.runGit(
			["diff", "--binary", "HEAD"],
			`Could not capture git diff in ${this.workdir}`,
		);
	}

	async hasChanges(): Promise<boolean> {
		const result = await this.coreGit.hasUncommittedChangesUnder({
			cwd: this.workdir,
			relativePath: ".",
		});
		return this.unwrapGitResult(result, `Could not read git status in ${this.workdir}`);
	}

	async createResultBranchAndCommit(branch: string, message: string): Promise<void> {
		const branchResult = await this.coreGit.createBranchAtHead({ cwd: this.workdir, branch });
		this.unwrapGitOperationResult(branchResult, `Could not create result branch ${branch}`);
		await this.runGit(["switch", branch], `Could not switch to result branch ${branch}`);
		await this.runGit(["add", "-A"], `Could not stage vibechk changes on ${branch}`);
		await this.runGit(["commit", "-m", message], `Could not commit vibechk changes on ${branch}`);
	}

	async checkout(branch: string): Promise<void> {
		await this.runGit(["switch", branch], `Could not switch back to branch ${branch}`);
	}

	private unwrapGitResult<T>(result: GitResult<T>, errorMessage: string): T {
		if (result.ok) return result.value;
		if (isMissingExecutableError(result.error.message)) {
			throw new VibechkError("git is not installed or not on PATH.");
		}
		throw new VibechkError(`${errorMessage}\n${result.error.message}`);
	}

	private unwrapGitOperationResult(result: GitOperationResult, errorMessage: string): void {
		if (result.ok) return;
		if (isMissingExecutableError(result.error.message)) {
			throw new VibechkError("git is not installed or not on PATH.");
		}
		throw new VibechkError(`${errorMessage}\n${result.error.message}`);
	}

	private async runGit(args: readonly string[], errorMessage: string): Promise<string> {
		const result = await this.runGitRaw(args);
		if (result.code !== 0 || result.killed) {
			throw new VibechkError(
				formatCommandFailure(errorMessage, formatCommand("git", args), result),
			);
		}
		return result.stdout.trim();
	}

	private async runGitRaw(args: readonly string[]): Promise<ExecResult> {
		let result: ExecResult;
		try {
			result = await this.execApi.exec("git", [...args], {
				cwd: this.workdir,
				timeout: GIT_TIMEOUT_MS,
			});
		} catch (error: unknown) {
			const message = error instanceof Error ? error.message : String(error);
			if (isMissingExecutableError(message)) {
				throw new VibechkError("git is not installed or not on PATH.");
			}
			throw new VibechkError(`git command failed before completion: ${message}`);
		}
		if (result.startupError !== undefined) {
			if (isMissingExecutableError(result.startupError)) {
				throw new VibechkError("git is not installed or not on PATH.");
			}
			throw new VibechkError(`git command failed before completion: ${result.startupError}`);
		}
		return result;
	}
}

function isMissingExecutableError(message: string): boolean {
	return message.includes("ENOENT");
}
